### DISCUSSION TITLE
Variable Verbraucher - virtuelle Python Wallbox

### DISCUSSION THREAD

**Discussion by BossoH** [camueller/SmartApplianceEnabler#214] [Q&A] [answered]
Hallo,
ich hatte in #197 schon mal eine Frage gehabt, wo es darum ging einen variablen Verbraucher mit dem SAE koppeln zu können. Wie es scheint geht das nur mit einer Wallbox. Meine Lösung ist nun eine Wallbox in Python nachzustellen. Falls das für jemanden interessant ist, habe ich anbei das Skript für meinen [Modbus Server (Slave)](https://github.com/camueller/SmartApplianceEnabler/files/8164303/pyModbusTCPServer.txt)
Hier wird also ein Modbus Server gestartet, der drei Register bereit stellt. Dies ist analog der Konfiguration einer echten Wallbox. In einem weiteren Programm (Client) muss dann: 
- der verfügbare Strom von diesem Server ausgelesen werden
- der Zustand gesetzt werden

In meinem Fall steuere ich mit diesen Informationen dann die GPIOs an.

Das Funktioniert prinzipiell super, aber wenn ich im Sunnyportal meinen Verbraucher ansehe, sehe ich den angefügten Verlauf.

![Unbenannt](https://user-images.githubusercontent.com/89273714/156237459-bbf5a2f0-181e-4052-b608-1b84592b1233.JPG)

Hat jemand eine Idee, warum die Leistung immer wieder regelmäßig auf 0 geht? Der SAE setzt in dieser Zeit immer den Wert "Ladevorgang beenden". Die überschüssige Leistung änderte sich in den aufgezeichneten 30 Minuten nur minimal. Es gibt auch keinen anderen Verbraucher, der diesen Zyklus erklären würde.

Viele Grüße

**Comment by camueller** [comment]
Coole Idee mit der virtuellen Wallbox!

Wenn der SHM immer einen Ausschaltbefehl schickt, dann meldet der SAE vermutlich keinen Bedarf an den SHM (prüfen an der SEMP-Schnittstelle des SAE). Bei Wallboxen wird automatisch Bedarf an Überschußenergie an den SHM gemeldet, sobald das Auto verbunden ist. Deine virtuelle Wallbox muss also dem SAE melden, dass das Fahrzeug verbunden ist.

**Comment by BossoH** [comment]
Das ist gegeben.

Ich habe folgende Schaltbedingungen eingestellt:
- immer 7 Uhr _Ladestecker angeschlossen_ 
- immer 20 Uhr _Ladestecker nicht angeschlossen_
- wenn "die Wallbox" vom SHM/SAE _Ladevorgang beginnen_ empfängt wird _Ladevorgang läuf_ gesetzt
- wenn "die Wallbox" vom SHM/SAE _Ladevorgang beenden_ empfängt wird _Ladestecker angeschlossen_ gesetzt

Um nicht in die Gefahr zu geraten, dauernd die Relaiskarte zu schalten, schaut mein Skript nur alle 30s nach den Modbus Registern. Könnte das zu selten sein? Gibt es da irgendwelche kritischen Zeitintervalle?

**Comment by camueller** [comment]
Der SAE fragt ja auch nur alle 20s die Wallbox ab, das passt also.
Wie schon geschrieben: Du must prüfen, ob der SAE dem SHM überhaupt einen Bedarf meldet. Wenn das Gerät eingeschaltet ist, aber kein Bedarf vorliegt und das Gerät Befehle akzeptiert, wird der SHM einen Ausschaltbefehl schicken. Alle relevanten Informationen findest Du im Log und /semp.

**Comment by BossoH** [comment, marked as answer]
Mein letzter Post ist lange her. Ich konnte mich leider eine Weile nicht um das Problem kümmern. Es scheint sich aber von alleine erledigt zu haben. Keine Ahnung, was das Problem war. Ich hoffe es kommt nicht wieder ;-)
